diff -aurN gamin-0.1.5.orig/server/local_inotify_syscalls.h gamin-0.1.5/server/local_inotify_syscalls.h
--- gamin-0.1.5.orig/server/local_inotify_syscalls.h	2005-08-09 16:56:43.000000000 +0200
+++ gamin-0.1.5/server/local_inotify_syscalls.h	2005-08-14 11:11:38.000000000 +0200
@@ -35,7 +35,7 @@
 # warning "Unsupported architecture"
 #endif
 
-#if defined(__i386__) || defined(__x86_64) || defined(__alpha__) || defined(__ppc__) || defined(__sparc__) || defined(__powerpc__) || defined(__powerpc64__) || defined(__ia64__) || defined(__s390__)
+#if defined(__i386__) || defined(__x86_64__) || defined(__alpha__) || defined(__ppc__) || defined(__sparc__) || defined(__powerpc__) || defined(__powerpc64__) || defined(__ia64__) || defined(__s390__)
 static inline int inotify_init (void)
 {
 	return syscall (__NR_inotify_init);
